tasks/task_manager.cc

```diff
@@ -36,11 +36,46 @@ task_manager::task::impl::impl(module_ptr module, task_id id, uint64_t sequence_
     }
 }
 
+future<std::optional<double>> task_manager::task::impl::expected_total_workload() const {
+    return make_ready_future<std::optional<double>>(std::nullopt);
+}
+
+std::optional<double> task_manager::task::impl::expected_children_number() const {
+    return std::nullopt;
+}
+
+task_manager::task::progress task_manager::task::impl::get_binary_progress() const {
+    return tasks::task_manager::task::progress{
+        .completed = is_finished(),
+        .total = 1.0
+    };
+}
+
 future<task_manager::task::progress> task_manager::task::impl::get_progress() const {
-    if (!_children.empty()) {
-        co_return progress{};
+    auto children_num = _children.size();
+    if (children_num == 0) {
+        co_return get_binary_progress();
+    }
+
+    std::optional<double> expected_workload = std::nullopt;
+    auto expected_children_num = expected_children_number();
+    // When get_progress is called, the task can have some of its children unregistered yet.
+    // Then if total workload is not known, progress obtained from children may be deceiving.
+    // In such a situation it's safer to return binary progress value.
+    if ((!expected_children_num || expected_children_num.value() != children_num) && !(expected_workload = co_await expected_total_workload())) {
```

Comment:
Can be simplified by `expected_children_num.value_or(0) != children_num` since `children_num != 0` at this point.